commands/src/main/java/discord4j/commands/CommandProvider.java

```diff
@@ -0,0 +1,20 @@
+package discord4j.commands;
+
+import discord4j.core.event.domain.message.MessageCreateEvent;
+
+import java.util.Optional;
+
+/**
+ * This represents a generic command provider.
+ */
+@FunctionalInterface
+public interface CommandProvider {
+
+    /**
+     * This is called to provide a matching command based on the event context.
+     *
+     * @param context The raw event context.
+     * @return The matched command based on the context, or empty if no command matched.
+     */
+    Optional<? extends BaseCommand> provide(MessageCreateEvent context);
```

Comment:
Would prefer Mono. I can think of a situation where a provider would lookup a suitable command to execute based off some parameters that are configured on a database.